### PR TITLE
[JIT] Make typing understand exceptions

### PR DESCRIPTION
**Pull request by eellison**
When we're emitting an if node, if one branch exits allow variables in the other branch to escape scope. This is using the same machinery that already exists for early returns so there are minimal changes to the compiler. Most of the changes are in the exit_transform pass so we don't create terrible graphs when exceptions exist. In a follow up PR i will add a writeup of the transform pass to docs since this should be the last change made to it for a while.

This will allow assertions to refine Optional types, as well as allow JIT to understand things like: 
```
def foo(x):
    if x == 1:
        raise Exception()
    else:
        a = 1
    return a
```

If you look in nn/functional.py, like 3/4 of the TODOs are this issue. One note is that if a function always throws, I accepted whatever the annotation for the return type is if it exists and otherwise set it to None. This is consistent with what mypy does.

### CODE REVIEW

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 1]
```diff
@@ -419,12 +430,25 @@ struct ExitTransformer {
     // exit values. since the exit does not extend outside this block,
     // update returned exit to false. then, reset the target_block to whatever
     // it was previously
-    // this will occur with a Loop block when transforming LoopContinuations
-    // and a Graph/Closure block when transforming ReturnStmts.
     if (target_block_ == block) {
-      TORCH_INTERNAL_ASSERT(getExitStatus(exit_pair) == ExitStatus::WILL);
-      registerBlockOutputs(block, exit_pair.exitValues());
-      exit_pair = ExitPair(false_val_, std::vector<Value*>({}));
+      if (getExitStatus(exit_pair) == ExitStatus::MIGHT) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: Could you please elaborate a bit on what's going on here?

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 2]
```diff
@@ -419,12 +430,25 @@ struct ExitTransformer {
...
+        exit_pair = constructWillExitPair(new_if->outputs());
+      }
+
+      if (getExitStatus(exit_pair) == ExitStatus::WILL) {
+        removeOutputs(block);
+        registerBlockOutputs(block, exit_pair.exitValues());
+      }
+
+      // reset the exiting status so that e.g. an exception from a closure block
+      // wont propagate to the enclosing function
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: Why don't we want to propagate exceptions up? (Sorry if you've explained to me and I forgot - if that's the case, then it's a good indicator that it deserves a comment :) )

**Review by ZolotukhinM — COMMENTED** [review]
I have several minor remarks and a couple of actual questions, please see inline.

File: torch/csrc/jit/script/exit_transforms.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/script/exit_transforms.cpp]: added comment inline

File: torch/csrc/jit/script/exit_transforms.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/script/exit_transforms.cpp]: added comment

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 3]
```diff
@@ -166,8 +158,10 @@ struct ExitTransformer {
     Block* body = loop.bodyBlock();
     auto exit_pair = transformExits(body);
     // if we're not exiting to outside the loop we don't need to do any work.
-    if (getExitStatus(exit_pair) == ExitStatus::WONT) {
-      return exit_pair;
+    // since we may not enter the loop return WONT for the THROWS case
```
[the remark this answers is not in the document]
Comment (ZolotukhinM) [reply torch/csrc/jit/script/exit_transforms.cpp]: Trying to summarize our offline discussion here: originally I was concerned with how we pick `WONT` even though the loop *might* throw, but actually `WONT` here is a more conservative choice than `THROWS` (we keep some outputs that never actually got used and add guarding conditions), so picking it is safe. The only suggestion from me would be to document that in a comment here, near the enum, or in the doc you plan to write.

**Review by ZolotukhinM — APPROVED** [review]
Thanks for addressing the remarks and clarifying the logic more!

**Pull request closed without merging**